Overhaul cudf Python developer docs (#22689)
rapidsai/cudf: docs/cudf/source/cudf/developer_guide/library_design.md

> ```{image} frame_class_diagram.png

Reviewer: Huh I was going to ask you to regenerate this figure but I guess it's somehow not that out of date? It looks like we updated it for pandas 2.0 when we reshuffled all the index classes.

Author: Yeah IIRC I changed this inheritance structure a while back and probably updated these diagram too

```

The class diagram shows the inheritance between various subclasses of the `Frame` which leads to public objects
like the `DataFrame`, `Series` and `Index`.

### Frames

The `Frame` is a base class that holds a data structure called the `ColumnAccessor` that maps one or more columns to a unique label.
The `Frame` and its subclass hierarchy is designed to consolidate shared implementation of private and public methods
that are applicable across the public `Series`, `DataFrame` and `Index` objects:

- `Frame` implements base methods that are common across `Series`, `DataFrame`, and `Index`.
- `SingleColumnFrame` implements methods that can be shared across `Series` and `Index` which are only represented by 1 column.

> - `IndexedFrame` implements methods that can be shared across `DataFrame` and `Series` which operates on the `.index`

Reviewer: ```suggestion
- `IndexedFrame` implements methods that can be shared across `DataFrame` and `Series`, both of which have an Index
```

## The Column layer

The next layer in the cuDF stack is the Column layer.
This layer forms the glue between pandas-like APIs and our underlying data layouts.
The principal objects in the Column layer are the `ColumnAccessor` and the various `Column` classes.
The `Column` is cuDF's core data structure that represents a single column of data of a specific data type.
A `ColumnAccessor` is a mapping of column labels to `Column`s.
A `Frame` owns a `ColumnAccessor`.

> ### ColumnAccessor

Reviewer: Are you sure we should drop this? ColumnAccessor remains a sort of hidden part of the library that is quite central to many operations. I wish it weren't, and maybe one day we'll refactor into a cleaner design, but for now it seems important to me to document. I'm open to being convinced otherwise.

Author: I moved it's mention in the `Frames` section as an object that `Frame` holds which is the mapping of `ColumnBase`s

IMO most of `ColumnAccessor` value is in facilitating DataFrame column indexing operations and is rarely a place I would consider a dev (agent) to implement something so that's why I excluded it.

methods are generally implemented using pylibcudf APIs and contains data type resolution logic to return a new `ColumnBase` of appropriate
data type if applicable.

`ColumnBase` has various subclasses eventually used by the `Frame`
that consolidates logic and implements methods specific to 1 or more related data types:

- `NumericalColumn`: integer, float, and boolean
- `StringColumn`: string
- `CategoricalColumn`: category
- `DatetimeColumn`: timestamp
- `DatetimeTZColumn`: timestamp with timezone
- `TimedeltaColumn`: duration
- `IntervalColumn`: interval
- `ListColumn`: list
- `StructColumn`: struct
- `Decimal32Column`: decimal32
- `Decimal64Column`: decimal64
- `Decimal128Column`: decimal128

Each column subclass is restricted to hold a `.dtype` object corresponding to a valid [pandas data type](https://pandas.pydata.org/docs/user_guide/basics.html#dtypes) of the same data type designation with the following exceptions:

- A `CategoricalColumn` holds a `cudf.CategoricalDtype` instead of a `pandas.CategoricalDtype`

> - A `IntervalColumn` can hold a `cudf.IntervalDtype` instead of a `pandas.IntervalDtype`

Reviewer: Is this right? CategoricalColumn _always_ uses the cudf dtype, whereas IntervalColumn might use either the cudf or the pandas type?

Author: >  CategoricalColumn always uses the cudf dtype

This is correct.

> IntervalColumn might use either the cudf or the pandas type

`IntervalColumn` can use `pd.ArrowDtype(IntervalArrowType)` but not `pandas.IntervalDtype`

(FWIW, I don't see a great reason anymore why we need a distinct cudf version of `IntervalDtype` and `CategoricalDtype)
